MantidPlot/src/ProjectSerialiser.cpp

```diff
@@ -0,0 +1,946 @@
+
+  if (!isTopLevel) {
+    text += saveFolderHeader(folder, isCurrentFolder);
+  }
+
+  text += saveFolderSubWindows(folder);
+
+  if (!isTopLevel) {
+    text += saveFolderFooter();
+  }
+
+  return text;
+}
+
+/**
+ * Generate the opening tags and meta information about
+ * a folder record for the Mantid project file.
+ *
+ * @param isCurrentFolder :: whether this folder is the current one.
+ * @return string representation of the folder's header data
+ */
+QString ProjectSerialiser::saveFolderHeader(Folder *folder,
+                                            bool isCurrentFolder) {
+  QString text;
+
+  // Write the folder opening tag
+  text += "<folder>\t" + QString(folder->objectName()) + "\t" +
+          folder->birthDate() + "\t" + folder->modificationDate();
+
+  // label it as current if necessary
+  if (isCurrentFolder) {
+    text += "\tcurrent";
+  }
+
+  text += "\n";
+  text += "<open>" + QString::number(folder->folderListItem()->isExpanded()) +
+          "</open>\n";
+  return text;
+}
+
+/**
+ * Generate the subfolder and subwindow records for the current folder.
+ * This method will recursively convert subfolders to their text representation
+ *
+ * @param app :: the current application window instance
+ * @param folder :: the folder to generate the text for.
+ * @param windowCount :: count of the number of windows
+ * @return string representation of the folder's subfolders
+ */
+QString ProjectSerialiser::saveFolderSubWindows(Folder *folder) {
+  QString text;
+
+  // Write windows
+  QList<MdiSubWindow *> windows = folder->windowsList();
+  for (auto w : windows) {
+    Mantid::IProjectSerialisable *ips =
+        dynamic_cast<Mantid::IProjectSerialisable *>(w);
+
+    if (ips) {
+      text += QString::fromUtf8(ips->saveToProject(window).c_str());
+    }
+  }
+
+  m_windowCount += windows.size();
+
+  // Write subfolders
+  QList<Folder *> subfolders = folder->folders();
+  foreach (Folder *f, subfolders) { text += saveFolderState(f); }
+
+  // Write log info
+  if (!folder->logInfo().isEmpty()) {
+    text += "<log>\n" + folder->logInfo() + "</log>\n";
+  }
+
+  return text;
+}
+
+/**
+ * Generate the closing folder data and end tag.
+ * @return footer string for this folder
+ */
+QString ProjectSerialiser::saveFolderFooter() { return "</folder>\n"; }
+
+/** This method saves the currently loaded workspaces in
+ * the project.
+ *
+ * Saves the names of all the workspaces loaded into mantid workspace tree.
+ * Creates a string and calls save nexus on each workspace to save the data
+ * to a nexus file.
+ *
+ * @return workspace names formatted as a Mantid project file string
+ */
+QString ProjectSerialiser::saveWorkspaces() {
+  using namespace Mantid::API;
+  std::string workingDir = window->workingDir.toStdString();
+  QString wsNames;
+  wsNames = "<mantidworkspaces>\n";
+  wsNames += "WorkspaceNames";
+
+  auto workspaceItems = AnalysisDataService::Instance().getObjectNames();
```

Comment:
Similarly, if `getObjectNames()` returned a reference (as some of Mantid's funtions do, but this one fortunately doesn't), then you would have to specify `auto& workspaecItems` to actually get the reference.

This is something to be aware of and can actually cause subtle bugs and performance hits.